source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+/*
+* Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+// For more information on how the RH hash works and in particular how we do
+// deletions, see:
+// http://codecapsule.com/2013/11/17/robin-hood-hashing-backward-shift-deletion/
+
+#include <aws/common/hash_table.h>
+#include <aws/common/common.h>
+#include <aws/common/math.h>
+#include <aws/common/linked_list.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <limits.h>
+
+// Include lookup3.c so we can (potentially) inline it and make use of the mix() macro.
+#include "private/lookup3.c"
+
+static void suppress_unused_lookup3_func_warnings() {
+    // We avoid making changes to lookup3 if we can avoid it, but since it has functions
+    // we're not using, reference them somewhere to suppress the unused function warning.
+    (void)hashword;
+    (void)hashword2;
+    (void)hashlittle;
+    (void)hashbig;
+}
+
+struct hash_table_entry {
+    struct aws_common_hash_element element;
+    uint64_t hash_code; // hash code (0 signals empty)
+};
+
+struct hash_table_state {
```

Comment:
AWS_CACHE_ALIGN ?

Reply:
This is a heap-allocated structure, adding an alignment annotation is a promise to the compiler, but not something that impacts the allocation directly.

It might be useful to align `slots` I suppose, but if we make that promise to the compiler then we need to have an aligned allocation function to use to allocate the state table.

Reply:
We probable need one of those anyways